mlflow/johnsnowlabs.py

```diff
@@ -0,0 +1,874 @@
+            code_paths=code_paths,
+            dfs_tmpdir=dfs_tmpdir,
+            sample_input=sample_input,
+            registered_model_name=registered_model_name,
+            signature=signature,
+            input_example=input_example,
+            await_registration_for=await_registration_for,
+            pip_requirements=pip_requirements,
+            extra_pip_requirements=extra_pip_requirements,
+            metadata=metadata,
+        )
+    # Otherwise, override the default model log behavior and save model directly to artifact repo
+    mlflow_model = Model(artifact_path=artifact_path, run_id=run_id)
+    with TempDir() as tmp:
+        tmp_model_metadata_dir = tmp.path()
+        _save_model_metadata(
+            tmp_model_metadata_dir,
+            spark_model,
+            mlflow_model,
+            sample_input,
+            conda_env,
+            code_paths,
+            signature=signature,
+            input_example=input_example,
+            pip_requirements=pip_requirements,
+            extra_pip_requirements=extra_pip_requirements,
+            remote_model_path=remote_model_path,
+        )
+        mlflow.tracking.fluent.log_artifacts(tmp_model_metadata_dir, artifact_path)
+        mlflow.tracking.fluent._record_logged_model(mlflow_model)
+        if registered_model_name is not None:
+            mlflow.register_model(
+                f"runs:/{run_id}/{artifact_path}",
+                registered_model_name,
+                await_registration_for,
+            )
+        return mlflow_model.get_model_info()
+
+
+def _save_model_metadata(
+        dst_dir,
+        spark_model,
+        mlflow_model,
+        sample_input,
+        conda_env,
+        code_paths,
+        signature=None,
+        input_example=None,
+        pip_requirements=None,
+        extra_pip_requirements=None,
+        remote_model_path=None,
+):
+    """
+    Saves model metadata into the passed-in directory.
+    If mlflowdbfs is not used, the persisted metadata assumes that a model can be
+    loaded from a relative path to the metadata file (currently hard-coded to "sparkml").
+    If mlflowdbfs is used, remote_model_path should be provided, and the model needs to
+    be loaded from the remote_model_path.
+    """
+    import pyspark
+
+    if sample_input is not None:
+        mleap.add_to_model(
+            mlflow_model=mlflow_model,
+            path=dst_dir,
+            spark_model=spark_model,
+            sample_input=sample_input,
+        )
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, dst_dir)
+
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, dst_dir)
+
+    # add the johnsnowlabs flavor
+    # TODO dynamic pyspark version # pyspark.__version__p
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        pyspark_version=johnsnowlabs.settings.raw_version_pyspark,
```

Comment:
Should this save the current version of PySpark that the model was saved with?

Reply:
okay now we are using `pysppark.__version__` . 
This means though for now the pip dependencies are updated to `johnsnowlabs_for_databricks` instead of `johnsnowlabs` . 

The only difference between `johnsnowlabs` and `johnsnowlabs_for_databricks` is that the latter one does not have pyspark as a dependency, giving the user flexibility to choose his own sparkversion. 


This constraint was originally put in place, because `johnsnowlabs.settings.raw_version_pyspark` defines the most stable version, in other versions some models may be broken. We are working on fixing that tough.